monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import java.io.InputStream
+
+import monix.eval.Task
+import monix.execution.Ack.Continue
+import monix.execution.{Ack, AsyncVar, Cancelable, Scheduler}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+
+class InputStreamResource(observable: Observable[Array[Byte]]) {
+
+  private type Bytes = Array[Byte]
+
+  def toTask: Task[InputStream] = {
+    Task.create[InputStream] { (s, cb) =>
+      val (state, cancelable) = subscribe(s)
+      cb.onSuccess(new ObservableInputStream(state, cancelable))
+    }
+  }
+
+  private def subscribe(s: Scheduler): (AsyncVar[Option[Bytes]], Cancelable) = {
+    val state = AsyncVar.empty[Option[Bytes]]()
+
+    val cancelable = observable.executeAsync.unsafeSubscribeFn(new Subscriber.Sync[Array[Byte]] {
+      implicit val scheduler = s
+
+      override def onNext(elem: Array[Byte]): Ack = {
+        state.put(Some(elem)).value
```

Comment:
it looks like `state.put(Some(elem)).map(_ => Continue)` causes `onComplete` to be delayed even if the stream (and consequently Observable) is cancelled. I will try to figure this out.

Reply:
Ah, right! 
Let me know if you get stuck

Reply:
@Avasil I am really a little stuck.There is a test case:
```
  testAsync("InputStream waits until enough bytes are available") {
    val array1 = randomByteArrayOfSize(10)
    val array2 = randomByteArrayOfSize(5)
    val array3 = randomByteArrayOfSize(5)
    var completed = false
    val observable = Observable.fromIterable(Seq(array1, array2, array3))
      .doOnComplete(Task {
        completed = true
      })

    Observable.toInputStream(observable)
      .map { inputStream =>
        val result = new Array[Byte](20)
        inputStream.read(result)

        assertEquals(result.toList, (array1 ++ array2 ++ array3).toList)
        assertEquals(inputStream.read(), -1.toByte)
        assert(completed)
      }
      .runToFuture(monix.execution.Scheduler.global)
  }
```
- I create a stream of 3 arrays
- I am trying to read 20 bytes
- and 2 out of 3 arrays come to `onNext`
- and both 2 are read in `ensureBufferSize`
- but then there are no `onNext` calls and the test hangs

This is visible in the ci test run https://travis-ci.org/monix/monix/builds/609307795?utm_source=github_status&utm_medium=notification

I tryied checking threads and they seem to be different in the `onNext`/`ensureBufferSize` so the execution has to continue, as I understand.
I also tried moving to built-in operators, because it looks like SafeSubscriber manages cancellations properly and I hoped that other operators at least handle concurrency properly. But it does not help either.

Could you please suggest how to debug this, because I do not where the next `onNext` call is stuck? 
Thank you.

Reply:
I don't know why the test has failed, it passed on CI after restart. I will try few more times to see if it comes back.

Does it also happen locally?

Reply:
yes, this happens locally

Reply:
How often does it happen? And do you run only this single test or all of them?
I'm trying to reproduce but so far it's passing on my machine

Reply:
Well, I can observe this once in several runs locally. 
I run it like this `sbt "testOnly monix.reactive.internal.builders.ToInputStreamSuite"`
There is the thread dump from the last run that stuck https://gist.github.com/melgenek/11ab38be0b1098cd732960f18a7ef92b

Reply:
I see, I will try again tomorrow and investigate

Reply:
I tried on different computer and I can finally reproduce it! I'll let you know when I figure it out

Reply:
BTW does it happen if we use `Scheduler.io` instead of `global`  or wrap `Await.result` in `blocking { }` ? 
If I add a timeout to `Await.result` then it seems like the next `onNext` continues after `inputStream.read()` times out and in one case I reproduced it happened on the same thread for some reason. Maybe it was the only thread available but that seems weird

Reply:
1) `blocking` did not help
2) I cannot reproduce the problem with `io` pool any more